AX: With ENABLE(AX_THREAD_TEXT_APIS), navigating by line to br elements no longer results in any VoiceOver speech or VoiceOver cursor

twilco opened the pull request:
#### 37aef476e4a0e82eb10684e0fe0121415d40a3c7
<pre>
AX: With ENABLE(AX_THREAD_TEXT_APIS), navigating by line to br elements no longer results in any VoiceOver speech or VoiceOver cursor
<a href="https://bugs.webkit.org/show_bug.cgi?id=291340">https://bugs.webkit.org/show_bug.cgi?id=291340</a>
<a href="https://rdar.apple.com/148945209">rdar://148945209</a>

Reviewed by Joshua Hoffman.

This commit fixes several bugs related to how we handle br elements, all regressing as a result of ENABLE(AX_THREAD_TEXT_APIS).

  - VoiceOver didn&apos;t speak anything when navigating to a br by line. This happened because creating a text marker via
    AXObjectCache::textMarkerDataForVisiblePosition early-exited for anything that wasn&apos;t a RenderText or replaced element.
    Fix this by handling br elements appropriately.

  - VoiceOver didn&apos;t draw a cursor for br elements when navigating to them by line. This was because we didn&apos;t cache
    a relative frame for them, which in turn happened because they are skipped by paint by default. Fix this by changing
    InlineContentPainter::paintDisplayBox to paint brs for PaintPhase::Accessibility, and add the appropriate handling
    in AccessibilityRegionContext.

  - VoiceOver read &quot;dimmed&quot; for br elements. This happened because we didn&apos;t cache AXProperty::IsEnabled for ignored
    objects. Thus when VoiceOver requested is-enabled, we returned the default value of false. Fix this by caching
    AXProperty::IsEnabled for all objects.

Also fix a crash in AXTextMarkerCocoa::toAttributedString where we tried to append a nil attributed string. This fix is
covered by new test, accessibility/mac/br-element-text-selection.html.

* LayoutTests/accessibility/mac/br-element-properties-expected.txt: Added.
* LayoutTests/accessibility/mac/br-element-properties.html: Added.
* LayoutTests/accessibility/mac/br-element-text-selection-expected.txt: Added.
* LayoutTests/accessibility/mac/br-element-text-selection.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations: Skip new tests.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/layout/integration/inline/LayoutIntegrationInlineContentPainter.cpp:
(WebCore::LayoutIntegration::InlineContentPainter::paintDisplayBox):
* Source/WebCore/rendering/AccessibilityRegionContext.cpp:
(WebCore::AccessibilityRegionContext::takeBounds):
* Source/WebCore/rendering/AccessibilityRegionContext.h:

Canonical link: <a href="https://commits.webkit.org/293518@main">https://commits.webkit.org/293518@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5738fa1d199ded093dd6b00ac7351188f25ecfbb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99139 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18782 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9028 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104259 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49720 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19071 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27219 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75446 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32563 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102144 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89511 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55812 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7487 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49096 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7573 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106623 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26248 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84424 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85715 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83923 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21293 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28583 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6257 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19961 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26189 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31370 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26010 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29323 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27576 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on LayoutTests/accessibility/mac/br-element-properties.html:
```diff
@@ -0,0 +1,63 @@
+<div id="editable" contenteditable="true">x<br/><br/><br/><br/>y</div>
+
+<script>
+var output = "This test ensures we can create valid text marker ranges from positions adjacent to replaced elements.\n\n";
```
I think this is leftover from a previous test. Can we update this to describe this test?

hoffmanjoshua commented on LayoutTests/accessibility/mac/br-element-properties.html:
```diff
@@ -0,0 +1,63 @@
+}
+</script>
+</body>
+</html>
```
These tags are duplicates of the ones above.

twilco committed:
AX: With ENABLE(AX_THREAD_TEXT_APIS), navigating by line to br elements no longer results in any VoiceOver speech or VoiceOver cursor
https://bugs.webkit.org/show_bug.cgi?id=291340
rdar://148945209

Reviewed by Joshua Hoffman.

This commit fixes several bugs related to how we handle br elements, all regressing as a result of ENABLE(AX_THREAD_TEXT_APIS).

  - VoiceOver didn't speak anything when navigating to a br by line. This happened because creating a text marker via
    AXObjectCache::textMarkerDataForVisiblePosition early-exited for anything that wasn't a RenderText or replaced element.
    Fix this by handling br elements appropriately.

  - VoiceOver didn't draw a cursor for br elements when navigating to them by line. This was because we didn't cache
    a relative frame for them, which in turn happened because they are skipped by paint by default. Fix this by changing
    InlineContentPainter::paintDisplayBox to paint brs for PaintPhase::Accessibility, and add the appropriate handling
    in AccessibilityRegionContext.

  - VoiceOver read "dimmed" for br elements. This happened because we didn't cache AXProperty::IsEnabled for ignored
    objects. Thus when VoiceOver requested is-enabled, we returned the default value of false. Fix this by caching
    AXProperty::IsEnabled for all objects.

Also fix a crash in AXTextMarkerCocoa::toAttributedString where we tried to append a nil attributed string. This fix is
covered by new test, accessibility/mac/br-element-text-selection.html.

* LayoutTests/accessibility/mac/br-element-properties-expected.txt: Added.
* LayoutTests/accessibility/mac/br-element-properties.html: Added.
* LayoutTests/accessibility/mac/br-element-text-selection-expected.txt: Added.
* LayoutTests/accessibility/mac/br-element-text-selection.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations: Skip new tests.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/layout/integration/inline/LayoutIntegrationInlineContentPainter.cpp:
(WebCore::LayoutIntegration::InlineContentPainter::paintDisplayBox):
* Source/WebCore/rendering/AccessibilityRegionContext.cpp:
(WebCore::AccessibilityRegionContext::takeBounds):
* Source/WebCore/rendering/AccessibilityRegionContext.h:

Canonical link: https://commits.webkit.org/293518@main
